src/routes/api/sections/bookSection.ts

```diff
@@ -0,0 +1,176 @@
+        // Finds schedule start and end for the day that the current hour falls on
+        const currWeekDay = currHourStart.weekday;
+        let scheduleDayStart = 0;
+        let scheduleDayEnd = 0;
+        for (const day of room.schedule) {
+            if (day.dayOfWeek == currWeekDay) {
+                scheduleDayStart = day.start;
+                scheduleDayEnd = day.end;
+            }
+        }
+
+        // Checks if the hour falls under an open time for the room and if it does adds hour to the list of available times
+        if (room.closed || currHourStart.weekday != currHourEnd.weekday || currHourStart.hour < scheduleDayStart || currHourEnd.hour > scheduleDayEnd) {
+            invalidHours = true;
+            break;
+        }
+
+        // Query a time block for the current hour
+        const bookingTimeBlock = await timeBlockModel
+            .findOne({
+                sectionId: sectionInformation._id,
+                startsAt: currHourStart.toJSDate(),
+                endsAt: currHourEnd.toJSDate(),
+            })
+            .catch((error) => {
+                res.status(500).json({ error });
+                return;
+            });
+
+        if (!bookingTimeBlock) {
+            // If a time block for the current hour does not exist
+            // Check if the amount of users will exceed to the section capacity
+            if (emails.length > sectionInformation.capacity) {
+                overCapacity = true;
+                break;
+            }
+        } else {
+            // If a time block for the current hour does exist
+            // Check if the amount of users already booked plus the new users will exceed to the section capacity
+            const newUsers = bookingTimeBlock.users.concat(emails);
+            if (newUsers.length > sectionInformation.capacity) {
+                overCapacity = true;
+                break;
+            }
+        }
+
+        // Increments hour start and end for next iteration
+        currHourStart = currHourStart.plus({ hours: 1 });
+        currHourEnd = currHourEnd.plus({ hours: 1 });
+    }
+
+    // If booking is not valid, respond with success as false and give reason as message
+    if (invalidHours) {
+        const response = {
+            sectionId: id,
+            success: false,
+            message: 'booking unsuccessful, times do not fall under open room hours',
+        };
+        res.status(200).json(response);
+        return;
+    }
+    if (overCapacity) {
+        const response = {
+            sectionId: id,
+            success: false,
+            message: 'booking unsuccessful, more users than available spots',
+        };
+        res.status(200).json(response);
+        return;
+    }
+
+    currHourStart = start;
+    currHourEnd = currHourStart.plus({ hours: 1 });
+
+    while (currHourStart < end) {
+        // Query a time block for the current hour
+        const bookingTimeBlock = await timeBlockModel
```

Comment:
Same comment about inefficient queries. You can probably just reuse an existing array of timeBlockModels that you'd get earlier